esp-hal/src/timer/mod.rs

```diff
@@ -1,6 +1,231 @@
 //! General-purpose timers.
 
+use fugit::{ExtU64, MicrosDurationU64};
+
 #[cfg(systimer)]
 pub mod systimer;
 #[cfg(any(timg0, timg1))]
 pub mod timg;
+
+/// Timer errors.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// The timer is already active.
+    TimerActive,
+    /// The timer is not currently active.
+    TimerInactive,
+    /// The alarm is not currently active.
+    AlarmInactive,
+}
+
+/// Functionality provided by any timer peripheral.
+pub trait Timer: crate::private::Sealed {
+    /// Start the timer.
+    fn start(&self);
+
+    /// Stop the timer.
+    fn stop(&self);
+
+    /// Reset the timer value to 0.
+    fn reset(&self);
+
+    /// Is the timer running?
+    fn is_running(&self) -> bool;
+
+    /// The current timer value.
+    fn now(&self) -> u64;
```

Comment:
I suppose we should change this to be https://docs.rs/fugit/latest/fugit/struct.Instant.html? Probably the 64bit variant, then we can compare instants between different timers.

Reply:
Should we be using `Instant` or `TimerInstantUXX`?

https://docs.rs/fugit/latest/fugit/type.TimerInstantU32.html
https://docs.rs/fugit/latest/fugit/type.TimerInstantU64.html